### PR TITLE
Make Triton work again

### PR DESCRIPTION
**Pull request by SzymonOzog**
still WIP but hope it's enough to get a lock, based on previous work at https://github.com/tinygrad/tinygrad/pull/855 - willing to split bounty with the author

### CODE REVIEW

**Comment by geohot** [comment]
Bounty locked, good luck!

**Comment by geohot** [comment]
How is this coming?

**Comment by SzymonOzog** [comment]
> 

all tests passing locally, just some cleaning up and will be ready for merge today/tommorow

**Comment by SzymonOzog** [comment]
update, my previous comment was terribly wrong I fixed a bunch of stuff and ran tests with cuda=1 getting cached from some of my previous runs which gave me wrong results. It still requires work but it's on the better side of it. Things still missing for certain are local defines, tests in CI(why is it giving me torch cuda errors?) and casts(although I'm not sure triton implements them by default, will reaserch further)

File: tinygrad/runtime/ops_triton.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+from typing import Any
+from tinygrad.ops import  Compiled
+from tinygrad.runtime.ops_cuda import RawCUDABuffer
+from tinygrad.codegen.linearizer import LinearizerOptions
+from tinygrad.renderer.triton import uops_to_triton
+from tinygrad.helpers import DEBUG
+from tinygrad.runtime.ops_cuda import pretty_ptx 
+
+
+class TritonProgram:
```
Comment (geohot) [inline tinygrad/runtime/ops_triton.py:14]: This should not be a runtime. Triton is a compiler.

Comment (SzymonOzog) [reply tinygrad/runtime/ops_triton.py:14]: Okay. Will merge with ops_cuda

**Comment by SzymonOzog** [comment]
update, most tests passing locally, one case of illegal memory access that I'm 99% sure I already know how to solve, one compiler error only for a specific case of `TestOps::test_output_padded_conv_transpose2d` that is a bit more problematic and getting the tests working on cpu. Hope to finish this week

**Comment by SzymonOzog** [comment]
All tests passing except TestSymbolicOps and TestSymbolicJit (https://github.com/tinygrad/tinygrad/issues/1678). Tests in CI still broken, will work on fixing them first and hop on to fixing Symbolic if nobody does till then. Can disable CI in this and continue in another PR if there is a rush to merge

**Comment by kuugu** [comment]
I think this needs to be handled in the triton renderer while generating the triton code from List[Uops]. 
var_vals will have a dictionary of the actual values associated with each of the Variable instances. 

One option seems to be loading the "resolved" shape values into the kernel as tl.constexprs

**Comment by SzymonOzog** [comment]
okay, since #1676 was not a bug, I added a fix for symbolic tests, all tests now passing locally

**Comment by geohot** [comment]
Oh it's looking good! If it's slow (I see it's timing out now), you are welcome to disable some tests.

**Comment by SzymonOzog** [comment]
okay, this is weird this always segfaults `TestEarlyExec::test_enable_early_exec` and freezes `TestSymbolicSymbolicOps::test_variable_expand_expr_none` both not triton related, maybe some memory leaks from ocelot? will continue looking into it but testing is not stable enough atm

**Comment by geohot** [comment]
See #1735 for why it's timing out.

You can disable the symbolic symbolic stuff for Triton if you want

File: tinygrad/renderer/triton.py [hunk 2]
```diff
@@ -90,13 +89,13 @@ def kk(s): kernel.append("  "*depth+s)
...
     elif uop == UOps.CONST or uop == UOps.DEFINE_ACC:
-        r[u] = ssa("const" if uop == UOps.CONST else "acc")
-        val = ('-' if math.isinf(args) and args<0 else'') + ('float("inf")' if math.isinf(args) else str(args))
-        if math.isnan(args): val = "float('nan')"
-        if len(local_size) > 0:
-          kk(f"{r[u]} = tl.full(({','.join([str(next_power_of_2(x)) for x in local_size])},),{val}, dtype={triton_dtypes[newvar]})") 
-        else:
-          kk(f"{r[u]} = tl.where(1, {val}, {val}).to({triton_dtypes[newvar]})")
+      r[u] = ssa("const" if uop == UOps.CONST else "acc")
+      val = ('-' if math.isinf(args) and args<0 else'') + ('float("inf")' if math.isinf(args) else str(args))
```
Comment (gijskoning) [inline tinygrad/renderer/triton.py:93]:
```suggestion
      val = (('-' if args<0 else '') + 'float("inf")') if math.isinf(args) else str(args)
```

**Comment by SzymonOzog** [comment]
Okay, this should be ready to merge now - had to disable some tests for ocelot - it seems to crash when triton is using shared memory, I'll look into the problem to try to find MRE and fix/report the bug in ocelot, but that might take time an I think will be better off in another PR so that triton get's CI for the period

File: tinygrad/renderer/triton.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  return '\n'.join([re.sub(r'\{\s*(%\w+)\s*\}', r'\1', line) for line in ptx_code.split('\n')])
+
+def define_scalar(local_size, triton_type, args):
+  val = (('-' if args<0 else '') + 'float("inf")') if math.isinf(args) else ('float("nan")' if math.isnan(args) else str(args))
+  if len(local_size) > 0:
+    return f"tl.full(({','.join([str(next_power_of_2(x)) for x in local_size])},),{val}, dtype={triton_type})"
+  return f"tl.where(1, {val}, {val}).to({triton_type})"
+
+def uops_to_triton(function_name:str, uops:List[UOp]):
+  kernel = []
```
Comment (gijskoning) [inline tinygrad/renderer/triton.py:36]: Can remove a lot of lines here to the like of hotz.
`kernel, bugs, signatures = [], [], [] ` etc...

File: tinygrad/renderer/triton.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  }
+  triton_dtypes = {dtypes.double: "tl.float64", dtypes.float32: "tl.float32", dtypes.float16: "tl.float16", dtypes.bool: "tl.int8", dtypes.int8: "tl.int8", dtypes.uint8: "tl.uint8", dtypes.int32: "tl.int32", dtypes.int64: "tl.int64"}
+  signature_dtypes = {dtypes.double: "*fp64",dtypes.float32: "*fp32", dtypes.float16: "*fp16", dtypes.bool: "*i8", dtypes.int8: "*i8", dtypes.uint8: "*u8", dtypes._arg_int32: "i32", dtypes.int32: "*i32", dtypes.int64: "*i64"}
+
+  for u in uops:
+    uop,newvar,vin,args,_ = u
+    if uop == UOps.LOOP:
+      for i,var in enumerate(args[0]):
+        if isinstance(var, NumNode): continue # python doesnt have block scope
+        else:
```
Comment (gijskoning) [inline tinygrad/renderer/triton.py:73]: Else can be removed since if block contains continue

File: tinygrad/renderer/triton.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  if isinstance(var.max, int): return var.max
+  return re.sub(r'\[(.*?)\]', '', str(var.max))[1:-1]
+
+#NOTE can be removed after https://github.com/gpuocelot/gpuocelot/issues/8 gets resolved
+def remove_single_scalar_curly_braces(ptx_code):
+  return '\n'.join([re.sub(r'\{\s*(%\w+)\s*\}', r'\1', line) for line in ptx_code.split('\n')])
+
+def define_scalar(local_size, triton_type, args):
+  val = (('-' if args<0 else '') + 'float("inf")') if math.isinf(args) else ('float("nan")' if math.isnan(args) else str(args))
+  if len(local_size) > 0:
```
Comment (gijskoning) [inline tinygrad/renderer/triton.py:31]: Move up in function for early quiting. Also put return on same line

Comment (SzymonOzog) [reply tinygrad/renderer/triton.py:31]: wdym by move up in function?

Comment (gijskoning) [reply tinygrad/renderer/triton.py:31]: Oh nvm messed up viewing it on phone, sorry!

File: tinygrad/renderer/triton.py (hunk 4 above)
Comment (SzymonOzog) [reply tinygrad/renderer/triton.py:77]: done

File: tinygrad/renderer/triton.py (hunk 3 above)
Comment (SzymonOzog) [reply tinygrad/renderer/triton.py]: done

**Comment by geohot** [comment]
Can you update after #1750?

**Comment by SzymonOzog** [comment]
all tests passing on gpu, ocelot broken(again) will look into it but it doesn't seem stable with triton compiler

**Comment by SzymonOzog** [comment]
`Tensor.eye(10).numpy()`
outputs
```
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]]
 ```
 when kernel is compiled from:
 ```
 import triton
import triton.language as tl
tl.core.TRITON_MAX_TENSOR_NUMEL = float('inf')
@triton.jit
def E_5_5_2_2(data0):
  gidx0 = tl.program_id(0) # 5
  gidx1 = tl.program_id(1) # 5
  lidx2 = tl.arange(0, 2)[:, None]
  lidx3 = tl.arange(0, 2)[None, :]
  tl.store(data0 + ((((gidx0*20)+(gidx1*2))+(lidx2*10))+lidx3), tl.where(((((((gidx0*9)+(gidx1*2))+(lidx2*10))+lidx3)%11)<1),1.0,0.0), mask = (((gidx0<5) and gidx1<5) and lidx2<2) and lidx3<2) 
  ```
  the program seems correct, and should definetly work especially for [0,0] but fails when ran in ocelot, will check what code was generated by last working version

**Comment by SzymonOzog** [comment]
the last working version generated
```
 import triton
 import triton.language as tl
 tl.core.TRITON_MAX_TENSOR_NUMEL = float('inf')
 @triton.jit
 def E_10_10(data0):
   gidx0 = tl.program_id(0) # 10
   gidx1 = tl.program_id(1) # 10
   alu0 = ((gidx0*10))
   alu1 = ((alu0+gidx1)).to(tl.int32)
   alu2 = ((alu1%t11)).to(tl.int32)
   alu3 = ((alu2<1)).to(tl.int8)
   alu4 = (tl.where(alu3, 1.0, 0.0)).to(tl.float32)
   tl.store(data0 + ((gidx0*10)+gidx1), alu4, mask = ((gidx0*10)<91) and gidx1<10) 
```
seems like the strides and global/local size generated by the uast changed but this shouldn't really be an issue

**Comment by SzymonOzog** [comment]
disabling local for ocelot helped, will look into what is wrong and open another PR when I get to something, ocelot seems to have bugs with how triton compiles local because using shared memory caused segfaults when ptx got passed to ocelot

**Comment by geohot** [comment]
Oh! I think I've seen this issue before. You have to tell CUDA about the shared memory, Triton needs it passed in. So I imagine you have to tell ocelet too. This may have been solved in one of the previous PRs.

**Comment by SzymonOzog** [comment]
I've checked and for the failing examples `prg.metadata["shared"] ==0`

**Comment by SzymonOzog** [comment]
update: shared is non 0 for the tests that segfaulted, I've found a way to fix this and made a pr https://github.com/gpuocelot/gpuocelot/pull/10 but that still doesn't fix broken tests that don't use shared like the Eye mentioned above, I'll look into it but that will prob be around monday as I'm out for the weekend

**Comment by SzymonOzog** [comment]
Found and fixed the bugs in ocelot that were breaking triton, currently switched gpuocelot to my branch, will switch back to master after https://github.com/gpuocelot/gpuocelot/pull/11 and https://github.com/gpuocelot/gpuocelot/pull/10 get accepted

File: tinygrad/codegen/optimizer.py [hunk 6]
```diff
@@ -131,7 +131,7 @@ def alias_buffer(self, i, pattern):
   def apply_auto_opt(self, x):
     for axis, amt, typ in x:
       if axis is None or amt == 1: continue
-      if typ == "G":
+      if typ == "G" and self.opts.has_shared:
```
Comment (geohot) [inline tinygrad/codegen/optimizer.py:134]: Do you still need this?

File: tinygrad/ops.py [hunk 7]
```diff
@@ -173,6 +173,8 @@ def __init__(self, buffer: Type[RawBuffer], linearizer_opts, renderer, runtime,
   def to_program(self, k):
     k.linearize()
     src = self.renderer(k.function_name, k.uops)
+    if len(src) == 4:
```
Comment (geohot) [inline tinygrad/ops.py:176]: You should be able to get the global/local sizes from the linearizer

File: tinygrad/shape/symbolic.py [hunk 8]
```diff
@@ -330,5 +330,5 @@ def sym_infer(a: Union[Node, int], var_vals: Dict[Variable, int]) -> int:
   ModNode: lambda self,ops,ctx: f"({self.a.render(ops,ctx)}%{self.b})",
   LtNode: lambda self,ops,ctx: f"({self.a.render(ops,ctx)}<{sym_render(self.b,ops,ctx)})",
   SumNode: lambda self,ops,ctx: f"({'+'.join(sorted([x.render(ops,ctx) for x in self.nodes]))})",
-  AndNode: lambda self,ops,ctx: f"({' and '.join(sorted([x.render(ops,ctx) for x in self.nodes]))})"
+  AndNode: lambda self,ops,ctx: ' and '.join(sorted([f"({x.render(ops, ctx)}" for x in self.nodes])) + len(self.nodes) * ')'
```
Comment (geohot) [inline tinygrad/shape/symbolic.py:333]: This shouldn’t be used anymore

File: tinygrad/ops.py (hunk 7 above)
Comment (SzymonOzog) [reply tinygrad/ops.py]: okay, I realized I don't need to invert global but local still needs to be overriden

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -240,47 +240,49 @@ jobs:
...
-          key: ubuntu22.04-gpuocelot-19626fc00b6ee321638c3111074269c69050e091
+          key: ubuntu22.04-gpuocelot-szymonozog-tinygrad
       - name: Clone/compile gpuocelot
-        if: (matrix.backend == 'cuda' || matrix.backend == 'ptx') && steps.cache-build.outputs.cache-hit != 'true'
+        if: (matrix.backend == 'cuda' || matrix.backend == 'ptx' || matrix.backend == 'triton') && steps.cache-build.outputs.cache-hit != 'true'
         run: |
-          git clone --recurse-submodules https://github.com/gpuocelot/gpuocelot.git ${{ github.workspace }}/gpuocelot
+          git clone --recurse-submodules --single-branch --branch tinygrad https://github.com/SzymonOzog/gpuocelot.git ${{ github.workspace }}/gpuocelot
           cd ${{ github.workspace }}/gpuocelot/ocelot
-          git checkout 19626fc00b6ee321638c3111074269c69050e091
```
Comment (geohot) [inline .github/workflows/test.yml:261]: How is the upstreaming going? I like that this pins a commit

Comment (SzymonOzog) [reply .github/workflows/test.yml:261]: the fixes still haven't been merged nor reviewed

**Comment by geohot** [comment]
If tests pass on rebased master, will marge and the bounty is yours!

**Comment by geohot** [comment]
Also: `DEBUG=4 FORWARD_ONLY=1 TRITON=1 python3 test/test_ops.py TestOps.test_big_gemm`

contains this line:

`tl.store(data0 + (((((gidx0*(tl.where(1, 8192, 8192).to(tl.int32)))+(gidx1*(tl.where(1, 64, 64).to(tl.int32))))+(lidx2*(tl.where(1, 1024, 1024).to(tl.int32))))+(lidx3*(tl.where(1, 4, 4).to(tl.int32))))+(tl.where(1, 514, 514).to(tl.int32))), acc10, mask = (((gidx0<8) and gidx1<4) 024).to(tl.int32))))+(lidx3*(tl.where(1, 4, 4).to(tl.int32))))+(tl.where(1, 259, 259).to(tl.int32))), acc7, mask = (((gidx0<8) and gidx1<4) anand lidx2<8) and lidx3<16) `

Why is `tl.where(1, 8192, 8192)` being emitted? What is that supposed to do?

**Comment by SzymonOzog** [comment]
1. Hmm, weird - seems like test helpers crashed it's not triton related, will look into it when I'm home
2. `tl.where(1, 8192, 8192).to(tl.int32)` is the only way I found to cast in triton, `tl.int32(8192)` wouldn't compile

**Comment by SzymonOzog** [comment]
Okay, I've sanity rerun tests 3 times and the segfault never happened again

**Comment by geohot** [comment]
Does just `*8192` work?

**Comment by SzymonOzog** [comment]
Nope, a bunch failing with compilation/assertion errors due to bad types

**Comment by geohot** [comment]
Congrats, bounty is yours! george@tinygrad.org to claim

**Comment by geohot** [comment]
Hmm, one of the tests is at 13 minutes now...if it's flaky I'll have to revert

File: .github/workflows/test.yml (hunk 9 above)
Comment (dmikushin) [reply .github/workflows/test.yml:261]: Sorry for delay, I've merged them now. Somehow, notifications miss my attention. Your work is very valuabe, so next time, if I miss something, please feel free to ping me by mentioning @dmikushin or a personal message.

Comment (dmikushin) [reply .github/workflows/test.yml:261]: Sorry for delay, I've merged them now. Somehow, notifications miss my attention. Your work is very valuabe, so next time, if I miss something, please feel free to ping me by mentioning @dmikushin or a personal message.

Comment (SzymonOzog) [reply .github/workflows/test.yml:261]: Thanks a lot for the merge, and the heads up. I definetly won't hesitate to ping in the future if we'll need a fast response

**Pull request merged**